src/transformers/trainer.py

```diff
@@ -818,17 +818,43 @@ def create_optimizer(self):
                     "weight_decay": 0.0,
                 },
             ]
-            optimizer_cls = Adafactor if self.args.adafactor else AdamW
-            if self.args.adafactor:
+            if self.args.adafactor and self.args.optim not in {"adamw_hf", "adafactor"}:
```

Comment:
this logic is unnecessary complicated, IMHO

1. deprecate --adafactor in favor of --optim adafactor
2. if `--adafactor` is passed set --optim adafactor

now you no longer need to ever consider `self.args.adafactor`